instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/EvReroute.kt

```diff
@@ -0,0 +1,219 @@
+                            "engine" to "electric",
+                            "ev_max_charge" to "8000",
+                            "ev_connector_types" to "tesla",
+                            "energy_consumption_curve" to "0,300;20,160;80,140;120,180",
+                            "ev_charging_curve" to "40000,70000",
+                        )
+                    )
+                    .build(),
+                object : NavigationRouterCallback {
+                    override fun onRoutesReady(
+                        routes: List<NavigationRoute>,
+                        routerOrigin: RouterOrigin
+                    ) {
+                        mapboxNavigation.setNavigationRoutes(routes)
+                    }
+
+                    override fun onFailure(
+                        reasons: List<RouterFailure>,
+                        routeOptions: RouteOptions
+                    ) {
+                        logE("onFailure reasons=$reasons", "DEBUG")
+                    }
+
+                    override fun onCanceled(
+                        routeOptions: RouteOptions,
+                        routerOrigin: RouterOrigin
+                    ) {
+                        logE("onCanceled", "DEBUG")
+                    }
+                }
+            )
+        }
+
+        // wait for tracking to start
+        mapboxHistoryTestRule.stopRecordingOnCrash("no location tracking") {
+            Espresso.onIdle()
+        }
+        locationTrackingIdlingResource.unregister()
+
+        // push off route location and wait for the off route event
+        offRouteIdlingResource.register()
+        runOnMainSync {
+            mockLocationReplayerRule.loopUpdate(offRouteLocationUpdate, times = 5)
+        }
+
+        mapboxHistoryTestRule.stopRecordingOnCrash("no off route") {
+            Espresso.onIdle()
+        }
+        offRouteIdlingResource.unregister()
+
+        // wait for tracking to start again
+        locationTrackingIdlingResource.register()
+        mapboxHistoryTestRule.stopRecordingOnCrash("no tracking") {
+            Espresso.onIdle()
+        }
+        locationTrackingIdlingResource.unregister()
+
+        val countDownLatch = CountDownLatch(1)
+        runOnMainSync {
+            mapboxNavigation.historyRecorder.stopRecording {
+                logE("history path=$it", "DEBUG")
+                countDownLatch.countDown()
+            }
+        }
+        countDownLatch.await()
+
+        // assert results
+        expectedStates.assert()
+
+        runOnMainSync {
+            val routeOptions = mapboxNavigation.getNavigationRoutes().first().routeOptions
+            val newWaypoints = routeOptions.coordinatesList()
+
+            check(newWaypoints.size == 2) {
+                "Expected 2 waypoints in the route after reroute but was ${newWaypoints.size}"
+            }
+            check(newWaypoints[1] == mockRoute.routeWaypoints.last())
+            with(routeOptions.unrecognizedJsonProperties!!){
```

Comment:
Is that the only EV specific logic that's being checked here? Why would you need a reroute for that? I mean you can check the same in the original route, can't you?
I thought NN would use all the waypoints for reroute request or something like that.
Right now I don't see any EV reroute specific checks here but maybe I don't understand something.

Reply:
it's the sanity test to verify that reroute is triggering and that every required property is requested.

Reply:
What's reroute to do with these properties? They are set when routes are requested and not changed afterwards, if I understand correctly.

Reply:
yes, it is. And that's what it's actually doing: check that all ev-related properties migrated to reroute request and new corresponding to a new route

Reply:
> check that all ev-related properties migrated to reroute request and new corresponding to a new route

But here you only check the new route's options, not request parameters. You can check them using `BaseMockRequestHandler#handledRequests`.

Reply:
I think it's debatable what can we check here: I'm checking that route options were refreshed and some options are kept. At the same time, I don't check that exact these values are used for requests because I can rely on other tests, that should check that.